Add PipeCD best practice blog 01 (#2995)
pipe-cd/pipecd: docs/content/en/blog/news/pipecd-best-practice-01.md

---
date: 2021-12-29
title: "PipeCD best practice 01 - operate your own PipeCD cluster"
linkTitle: "PipeCD best practice 01"

> weight: 997

Author: ```suggestion
weight: 996
```

### Again, review the PipeCD architecture

![images](/images/architecture-overview-with-roles.png)
<p style="text-align: center;">
Component Architecture
</p>

> Note: Please refer to [concepts](/docs/concepts/) docs for definitions of PipeCD components such as Control Plane, Piped, Application, etc.

At a glance, PipeCD is composed of 2 components, the Control Plane and the Piped(s). As the figure above, depending on your task/role, you need to work with different components of PipeCD.

1. As a operator - platform team, most of your time working with PipeCD is to operate the PipeCD Control Plane. You may also need to cooperate with the product team, helping them install Piped(s) to their applications running cluster depending on your company team's structure. The space surrounded by a <span style="color: green;">green border</span> shows the operators working area.

> 2. As a developer - product team, most of your time working with PipeCD is to control your applications' development via the PipeCD application configuration files. You may also need to cooperate with the platform team to install Piped(s) to your applications running cluster if necessary. The space surrounded by a <span style="color: blue;">blue border</span> shows the developers working area.

Reviewer: ```suggestion
2. As a developer - product team, most of your time working with PipeCD is to deploy, manage and observe your applications via application configuration files and PipeCD web console. You may also need to cooperate with the platform team to install Piped(s) to your applications running cluster if necessary. The space surrounded by a <span style="color: blue;">blue border</span> shows the developers working area.
```

### Tips to operate PipeCD cluster

> 1. Only a single Control Plane needs to be installed and will be operated by the operators (platform team). The installed Control Plane doesn't have to be in the same cluster as your applications running cluster or the Piped installed cluster. Just make sure the Piped(s) can connect to the Control Plane via outbound requests is enough.

Reviewer: 👍

> 2. Developers from product team may need to self install piped into their applications running cluster (due to some credential related issues, if platform do not the one who manage those credentials).\

Reviewer 2: I'm sorry if they're wrong. 🙇 
```suggestion
2. Developers from product team may need to self install piped into their applications running cluster (due to some credential related issues, if platform do not be the one who manage those credentials).\
```

Author: nice catch, thank you 🙏

> 2. Developers from the product team may need to self-install piped into their applications running cluster (due to some credential-related issues, if the platform team does not manage those credentials).\

Reviewer: Is this the same with the number 3?

Author: Not really, what I want to mention in this (2) is that piped can be installed by platform team or developer team, while in (3) I want to mention that `we highly recommend running each Piped inside each cluster and that Piped will only manage the applications on that cluster` instead of using single piped to handle everything.

Reviewer: > (due to some credential-related issues,

Somehow I want to avoid this expression, It could lead to a misunderstanding of our security issue.

Reviewer: @khanhtc1202 Can you find a better sentence for this and we can merge and publish this blog.

Author: Sorry I missed the above message. Will update to avoid that.

> 2. Developers from the product team may need to self-install piped into their applications running cluster (due to some credential-related issues, if the platform team does not manage those credentials).\

Author: ```suggestion
2. Developers from the product team may need to self-install piped into their applications running cluster if the platform team does not manage those credentials.\
```

\

> When using PipeCD, developers only need to care about [application configuration](/docs/user-guide/adding-an-application/) - which defines the pipeline so that piped(s) can use them to deploy your applications. Interaction with the applications is mainly done via the PipeCD web console and the configuration files stored on Git.

Author: ```suggestion
Once the Piped - the runner is installed successfully, developers only need to care about [application configuration](/docs/user-guide/adding-an-application/) - which defines the pipeline so that piped(s) can use them to deploy your applications. Interaction with the applications is mainly done via the PipeCD web console and the configuration files stored on Git.
```

> As an __operator__, PipeCD [control-plane](/docs/operator-manual/control-plane/) and [piped](/docs/operator-manual/piped/)(s) are what you should care about. As a __developer__, you should care about [piped](/docs/operator-manual/piped/) which installed in your applications running cluster and your [applications' PipeCD configurations](/docs/user-guide/adding-an-application/).

> 3. It is possible to use a single piped to manage deployments of all the applications to the clusters other than the one in which piped is being installed. However, due to credentials and security issues, __piped should be installed independently in each applications running cluster, and only manages the deployments of applications on that cluster__. In case you want to deploy applications that are required to be installed across clusters/regions/environments, and want to follow this rule of thumb, please refer to the [deployment-chain](/docs/user-guide/deployment-chain/) feature.

Reviewer: > However, due to credentials and security issues, __piped should be installed independently in each applications running cluster, and only manages the deployments of applications on that cluster__

However, as security best practices and avoid generating cluster credentials, we highly recommend running each Piped inside each cluster and that Piped will only manage the applications on that cluster.
